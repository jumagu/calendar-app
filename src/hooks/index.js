export * from "./useForm";
export * from "./useUiStore";
export * from "./useAuthStore";
export * from "./useCalendarStore";