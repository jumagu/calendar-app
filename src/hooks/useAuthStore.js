import { useDispatch, useSelector } from "react-redux";

import {
  checkingCredentials,
  clearErrorMessage,
  clearEventsLogout,
  login,
  logout,
} from "../store";

import { calendarApi } from "../apis";

export const useAuthStore = () => {
  const dispatch = useDispatch();

  const { status, user, errorMessage } = useSelector((state) => state.auth);

  const startLogin = async ({ email, password }) => {
    dispatch(checkingCredentials());

    try {
      const { data } = await calendarApi.post("/auth", { email, password });

      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ name: data.name, uid: data.uid }));
    } catch (error) {
      if (error.response && !error.response.data.ok) {
        dispatch(logout("Invalid Credentials"));
      } else {
        dispatch(logout("Internal Server Error"));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegistry = async ({ name, email, password }) => {
    dispatch(checkingCredentials());

    try {
      const { data } = await calendarApi.post("/auth/new", {
        name,
        email,
        password,
      });

      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ name: data.name, uid: data.uid }));
    } catch (error) {
      if (error.response && !error.response.data.ok) {
        dispatch(logout(error.response?.data.msg));
      } else {
        dispatch(logout("Internal Server Error"));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");

    if (!token) return dispatch(logout());

    try {
      const { data } = await calendarApi.get("auth/renew");

      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ name: data.name, uid: data.uid }));
    } catch (error) {
      localStorage.clear();
      dispatch(logout());
    }
  };

  const startLogout = () => {
    localStorage.clear();

    dispatch(logout());
    dispatch(clearEventsLogout());
  };

  return {
    // * Properties
    user,
    status,
    errorMessage,

    // * Methods
    startLogin,
    startLogout,
    startRegistry,
    checkAuthToken,
  };
};
