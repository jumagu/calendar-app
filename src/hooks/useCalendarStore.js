import { useSelector, useDispatch } from "react-redux";

import {
  setActiveEvent,
  addNewEvent,
  editEvent,
  deleteEvent,
  loadEvents,
  closeDateModal,
} from "../store";
import { calendarApi } from "../apis";
import { convertEventDate } from "../helpers";
import Swal from "sweetalert2";

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent, isLoadingEvents } = useSelector(
    (state) => state.calendar
  );
  const { user } = useSelector((state) => state.auth);

  const handleSetActiveEvent = (calendarEvent) => {
    dispatch(setActiveEvent(calendarEvent));
  };

  const startLoadingEvents = async () => {
    try {
      const { data } = await calendarApi.get("/events");

      const events = convertEventDate(data.events);

      dispatch(loadEvents(events));
    } catch (error) {
      console.log(error);
    }
  };

  const startSavingEvent = async (calendarEvent) => {
    try {
      if (calendarEvent.id) {
        // Upadating
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);

        dispatch(editEvent({ ...calendarEvent, user }));

        return;
      }
      // Creating
      const { data } = await calendarApi.post("/events", calendarEvent);

      dispatch(addNewEvent({ ...calendarEvent, id: data.event.id, user }));
    } catch (error) {
      console.log(error);
      Swal.fire("Error editing event", error.response.data?.msg, "error");
    }
  };

  const startDeletingEvent = async () => {
    try {
      await calendarApi.delete(`/events/${activeEvent.id}`);

      dispatch(closeDateModal());
      dispatch(deleteEvent());
    } catch (error) {
      console.log(error);
      Swal.fire("Error deleting event", error.response.data?.msg, "error");
    }
  };

  return {
    // * Properties
    events,
    activeEvent,
    isLoadingEvents,
    hasEventSelected: !!activeEvent,

    // * Methods
    startSavingEvent,
    startLoadingEvents,
    startDeletingEvent,
    handleSetActiveEvent,
  };
};
