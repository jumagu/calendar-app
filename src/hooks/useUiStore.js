import { useSelector, useDispatch } from "react-redux";

import { openDateModal, closeDateModal } from "../store";

export const useUiStore = () => {
  const dispatch = useDispatch();

  const { isDateModalOpen } = useSelector((state) => state.ui);

  const handleOpenDateModal = () => {
    dispatch(openDateModal());
  };

  const handleCloseDateModal = () => {
    dispatch(closeDateModal());
  };

  const toggleDateModal = () => {
    isDateModalOpen ? handleCloseDateModal() : handleOpenDateModal();
  };

  return {
    // * Properties
    isDateModalOpen,

    // * Methods
    handleOpenDateModal,
    handleCloseDateModal,
    toggleDateModal,
  };
};
