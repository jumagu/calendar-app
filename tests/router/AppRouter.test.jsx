/* eslint-disable no-undef */

import { render, screen } from "@testing-library/react";

import { MemoryRouter } from "react-router-dom";

import { AppRouter } from "../../src/router/AppRouter";
import { useAuthStore } from "../../src/hooks/useAuthStore";

jest.mock("../../src/hooks/useAuthStore");
jest.mock("../../src/calendar", () => ({
  CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe("Pruebas en <AppRouter />", () => {
  const mockCheckAuthToken = jest.fn();
  beforeEach(() => jest.clearAllMocks());

  test("Debe mostrar la pantalla de carga y llamar el checkAuthToken()", () => {
    useAuthStore.mockReturnValue({
      status: "checking",
      checkAuthToken: mockCheckAuthToken,
    });

    render(<AppRouter></AppRouter>);

    expect(mockCheckAuthToken).toHaveBeenCalled();
    expect(screen.getByText("Loading...")).toBeTruthy();
  });

  test("Debe mostrar la pantalla de login si no está autenticado", () => {
    useAuthStore.mockReturnValue({
      status: "not-authenticated",
      checkAuthToken: mockCheckAuthToken,
    });

    const { container } = render(
      <MemoryRouter initialEntries={["/auth2/algo/otracosa"]}>
        <AppRouter></AppRouter>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText("Login")).toBeTruthy();
  });

  test("Debe mostrar la pantalla de calendario si está autenticado", () => {
    useAuthStore.mockReturnValue({
      status: "authenticated",
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter>
        <AppRouter></AppRouter>
      </MemoryRouter>
    );

    screen.debug();
    expect(screen.getByText("CalendarPage")).toBeTruthy();
  });
});
