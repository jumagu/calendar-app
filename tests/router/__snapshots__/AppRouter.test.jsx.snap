// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> Debe mostrar la pantalla de login si no est√° autenticado 1`] = `
<div>
  <div
    class="d-flex justify-content-center align-items-center px-3 login-container"
  >
    <div
      class="d-sm-flex w-100"
    >
      <div
        class="login-form-1"
      >
        <h3>
          Login
        </h3>
        <form>
          <div
            class="form-floating mb-2"
          >
            <input
              class="form-control "
              name="loginEmail"
              placeholder="Your email"
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            />
            <label>
              Email address
            </label>
          </div>
          <div
            class="input-group has-validation mb-2"
          >
            <div
              class="form-floating is-invalid"
            >
              <input
                class="form-control "
                name="loginPassword"
                placeholder="Your password"
                type="password"
                value=""
              />
              <label>
                Password
              </label>
            </div>
            <button
              class="btn bg-light-subtle border-light-subtle"
              style="width: 50px;"
              type="button"
            >
              <i
                class="fa-regular fa-eye"
              />
            </button>
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="form-group mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Enter"
            />
          </div>
        </form>
      </div>
      <div
        class="login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-floating mb-2"
          >
            <input
              class="form-control "
              name="registerName"
              placeholder="Your name"
              type="text"
              value=""
            />
            <div
              class="invalid-feedback fw-semibold"
            />
            <label>
              Name
            </label>
          </div>
          <div
            class="form-floating mb-2"
          >
            <input
              class="form-control "
              name="registerEmail"
              placeholder="Your email"
              type="email"
              value=""
            />
            <div
              class="invalid-feedback fw-semibold"
            />
            <label>
              Email
            </label>
          </div>
          <div
            class="input-group has-validation mb-2"
          >
            <div
              class="form-floating is-invalid"
            >
              <input
                class="form-control "
                name="registerPassword"
                placeholder="Your password"
                type="password"
                value=""
              />
              <label>
                Password
              </label>
            </div>
            <button
              class="btn bg-light-subtle border-light-subtle"
              style="width: 50px;"
              type="button"
            >
              <i
                class="fa-regular fa-eye"
              />
            </button>
            <div
              class="invalid-feedback fw-semibold"
            />
          </div>
          <div
            class="form-group mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
